File: android-twa/app/src/main/java/app/polycal/widgets/WidgetUrls.java
package app.polycal.widgets;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;

/**
 * Builds same-origin compose deep-links. Widgets never submit events — they open
 * the TWA LauncherActivity so Chrome reuses the Auth.js session (PC-479).
 */
public final class WidgetUrls {
    public static final String MODE_EVENT = "event";
    public static final String MODE_NLP = "nlp";
    public static final String HOST_ORIGIN = "https://polycal.net";

    private static final int TITLE_MAX = 256;
    private static final int NLP_MAX = 1024;

    private WidgetUrls() {}

    /**
     * Builds /feed?compose= with optional title or q, capped so the query cannot bloat.
     */
    public static Uri composeUri(String mode, String prefill) {
        Uri.Builder builder = Uri.parse(HOST_ORIGIN + "/feed").buildUpon()
                .appendQueryParameter("compose", mode);
        String text = prefill == null ? "" : prefill.trim();
        if (!text.isEmpty()) {
            if (MODE_NLP.equals(mode)) {
                builder.appendQueryParameter("q", text.length() > NLP_MAX ? text.substring(0, NLP_MAX) : text);
            } else {
                builder.appendQueryParameter("title", text.length() > TITLE_MAX ? text.substring(0, TITLE_MAX) : text);
            }
        }
        return builder.build();
    }

    /**
     * Opens the compose URL inside this TWA package (not a Chrome Custom Tab).
     */
    public static void launchComposer(Activity activity, String mode, String prefill) {
        Uri uri = composeUri(mode, prefill);
        Intent intent = new Intent(Intent.ACTION_VIEW, uri);
        intent.setPackage(activity.getPackageName());
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
        try {
            activity.startActivity(intent);
        } catch (Exception ignored) {
            activity.startActivity(new Intent(Intent.ACTION_VIEW, uri));
        }
    }
}

File: android-widgets/app/src/main/java/app/polycal/widgets/WidgetUrls.kt
package app.polycal.widgets

import android.net.Uri
import androidx.browser.customtabs.CustomTabsIntent

/**
 * Builds same-origin compose deep-links the PWA already understands.
 * Widgets never submit events themselves — they only launch the web composer.
 */
object WidgetUrls {
    const val MODE_EVENT = "event"
    const val MODE_NLP = "nlp"

    private const val TITLE_MAX = 256
    private const val NLP_MAX = 1024

    /**
     * Normalizes a host URL to http(s) origin only so a bad gradle property
     * cannot launch javascript: or file: URIs.
     */
    fun sanitizeBaseUrl(raw: String): String {
        val trimmed = raw.trim().trimEnd('/')
        val uri = Uri.parse(trimmed)
        val scheme = uri.scheme?.lowercase()
        if (scheme != "https" && scheme != "http") {
            return BuildConfig.POLYCAL_BASE_URL.trim().trimEnd('/')
        }
        if (uri.host.isNullOrBlank()) {
            return BuildConfig.POLYCAL_BASE_URL.trim().trimEnd('/')
        }
        return trimmed
    }

    fun composeUri(baseUrl: String, mode: String, prefill: String): Uri {
        val builder = Uri.parse("${sanitizeBaseUrl(baseUrl)}/feed").buildUpon()
            .appendQueryParameter("compose", mode)
        val text = prefill.trim()
        if (text.isNotEmpty()) {
            if (mode == MODE_NLP) {
                builder.appendQueryParameter("q", text.take(NLP_MAX))
            } else {
                builder.appendQueryParameter("title", text.take(TITLE_MAX))
            }
        }
        return builder.build()
    }

    fun launchComposer(activity: android.app.Activity, mode: String, prefill: String) {
        val uri = composeUri(BuildConfig.POLYCAL_BASE_URL, mode, prefill)
        val customTabs = CustomTabsIntent.Builder()
            .setShowTitle(true)
            .setShareState(CustomTabsIntent.SHARE_STATE_OFF)
            .build()
        try {
            customTabs.launchUrl(activity, uri)
        } catch (_: Exception) {
            // Fallback when Chrome Custom Tabs is unavailable (rare OEM / no browser).
            activity.startActivity(
                android.content.Intent(android.content.Intent.ACTION_VIEW, uri),
            )
        }
    }
}
